examples/notebooks/get-started/learn_beam_basics_by_doing.ipynb

```diff
@@ -0,0 +1,1068 @@
+        "\n",
+        "outputs = (\n",
+        "    pipeline\n",
+        "      | 'Take in Dataset' >> beam.io.ReadFromText(inputs_pattern)\n",
+        "      | 'Separate to list' >> beam.Map(lambda line: line.split(\"\\t\"))\n",
+        "      | 'Keep only spam' >> beam.Filter(lambda line: line[0] == \"spam\")\n",
+        "      | 'Find words' >> beam.FlatMap(lambda line: re.findall(r\"[a-zA-Z']+\", line[1]))\n",
+        "      | 'Pair words with 1' >> beam.Map(lambda word: (word, 1))\n",
+        "      # ADDED\n",
+        "      | 'Group and sum' >> beam.CombinePerKey(sum)\n",
+        "      | 'Write results' >> beam.io.WriteToText(\"ansoutput4\", file_name_suffix = \".txt\")\n",
+        "      | 'Print the text file name' >> beam.Map(print)\n",
+        ")\n",
+        "\n",
+        "pipeline.run()\n",
+        "\n",
+        "! head ansoutput4*.txt"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "L9zNI4LMJHur"
+      },
+      "source": [
+        "And we finished! Now that we have a count of all the words in our data set, it makes it much easier for us to use this feature to create a logistic regression model to detect spam texts."
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "BozCqbzUPItB"
+      },
+      "source": [
+        "\n",
+        "\n",
+        "---\n",
+        "\n",
+        "\n",
+        "\n",
+        "---\n",
+        "\n"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "Cv9508k4MeS4"
+      },
+      "source": [
+        "# Full Spam Ham Apache Beam Example\n",
+        "\n",
+        "Below is a summary of all the code we performed for your convenience. Note you do not need to explicitly call run with the with statement."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "8iWWy3Y7Msxm"
+      },
+      "outputs": [],
+      "source": [
+        "!pip install --quiet apache-beam\n",
+        "!mkdir -p data\n",
+        "!gsutil cp gs://apachebeamdt/SMSSpamCollection data/"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "43oijI4BCAW1"
+      },
+      "outputs": [],
+      "source": [
+        "import apache_beam as beam\n",
+        "import re\n",
+        "\n",
+        "inputs_pattern = 'data/SMSSpamCollection'\n",
+        "outputs_prefix_ham = 'outputs/fullcodeham'\n",
+        "outputs_prefix_spam = 'outputs/fullcodespam'\n",
+        "\n",
+        "# Ham Word Count\n",
+        "with beam.Pipeline() as pipeline:\n",
+        "     ham = (\n",
+        "      pipeline\n",
+        "      | 'Take in Dataset' >> beam.io.ReadFromText(inputs_pattern)\n",
+        "      | 'Separate to list' >> beam.Map(lambda line: line.split(\"\\t\"))\n",
+        "      | 'Keep only ham' >> beam.Filter(lambda line: line[0] == \"ham\")\n",
```

Comment:
one possibility is to use tagged outputs to get both spam and ham after 1 pass, see: https://beam.apache.org/documentation/programming-guide/#multiple-outputs-dofn

Adds complexity though.

Reply:
let us keep this notebook simple for beginners.